lib/SILOptimizer/Mandatory/PhiStorageOptimizer.cpp

```diff
@@ -262,39 +293,100 @@ void PhiStorageOptimizer::tryCoalesceOperand(SILBasicBlock *incomingPred) {
   coalescedPhi.coalescedOperands.push_back(incomingOper);
 }
 
-// Record liveness generated by uses of \p incomingVal.
+/// The block which dominates all the defs whose storage is projected out of the
+/// storage for \p incomingVal.
+///
+/// Populates \p projectedValues with the values whose storage is recursively
+/// projected out of the storage for incomingVal.
+SILBasicBlock *
+PhiStorageOptimizer::getDominatingBlock(SILValue incomingVal,
+                                        StackList<SILValue> *projectedValues) {
+  assert(domInfo);
+
+  // Recursively find the set of value definitions and the dominating LCA.
+  ValueWorklist values(incomingVal->getFunction());
+
+  values.push(incomingVal);
+
+  SILBasicBlock *lca = incomingVal->getParentBlock();
+  auto updateLCA = [&](SILBasicBlock *other) {
+    lca = domInfo->findNearestCommonDominator(lca, other);
+  };
+
+  while (auto value = values.pop()) {
+    if (projectedValues)
+      projectedValues->push_back(value);
+    auto *defInst = value->getDefiningInstruction();
+    if (!defInst) {
+      assert(!PhiValue(value));
+      updateLCA(value->getParentBlock());
+      continue;
+    }
+    SmallVector<Operand *, 2> operands;
+    if (findUseProjections(defInst, operands)) {
+      // Any operand whose storage is a use-projection out of `value`'s storage
+      // dominates `value` (i.e. `defInst`), so skip updating the LCA here.
+      for (auto *operand : operands) {
+        values.pushIfNotVisited(operand->get());
+      }
+      continue;
+    }
+    SmallVector<SILValue, 2> projecteds;
+    if (findDefProjections(value, &projecteds)) {
+      // Walking up the projection hain from this point, every subsequent
+      // projection must be a def projection [projection_chain_structure].
+      // Every such projection is dominated by `value` (i.e. defInst).
+      //
+      // If the walk were only updating the LCA, it could stop here.
+      //
+      // It is also collecting values whose storage is projected out of the
+      // phi, however, so the walk must continue.
+      updateLCA(defInst->getParent());
```

Comment:
Can you assert that `projecteds` is non-empty here? If, so, why do you need to update the LCA given that the projecteds must also update the LCA?

Reply:
Yeah, we can assert that--`projecteds` is empty if and only `findDefProjections` returns false.  Asserted here and for findUseProjections.

I don't think we can skip updating the LCA here: it needs to dominate `defInst->getParent()`.  The structure of chain projections is such that in order for a def projection to appear, it has to look something like this (`projection_chain_structure`):

```
parent:
  %val = ...
  cond_br ..., left, right

left:
  %d = disaggregate %val
  ...

right:
  %a = aggregate %val

...

  br merge( aggregate(...(aggregate(%a)...) )
```

In this case, so far so good: we'd update the LCA with `%d = disaggregate %val` which would make `parent` the LCA.  If the branch that's the LCA of `disaggregate` and `aggregate` is different from the block containing the def `%val`, however, the LCA is after the def block:


```
grandparent:
  %val = ...
  cond_br ..., parent, other

parent:
  cond_br ..., left, right

left:
  %d = disaggregate %val
  ...

right:
  %a = aggregate %val

...

  br merge( aggregate(...(aggregate(%a)...) )
```

In the worst case, because the LCA is wrong, the walk where liveness is determined continues past the LCA (since we just compare the current block to the LCA before pushing its predecessors) resulting in redundant `alloc_stack`s.  Without including this `defInst->getParent()` in the LCA computation, PhiStorageOptimizer incorrectly determines that there's interference in `entry` because the walk back from `destroy_value %v` continues all the way to `entry` (because LCA is incorrectly determined to be `parent` rather than `trouble2`.

```
sil [ossa] @g : $@convention(thin) <T> () -> () {
entry:
  %agg4 = apply undef<Box<Box<T>>>() : $@convention(thin) <T> () -> (@out T)
  cond_br undef, okay, trouble

okay:
  br merge(%agg4 : $Box<Box<T>>)


trouble:
  destroy_value %agg4 : $Box<Box<T>>
  br trouble2

trouble2:
  %v = apply undef<Box<T>>() : $@convention(thin) <T> () -> (@out T)
  cond_br undef, grandparent, other

grandparent:
  br parent

parent:
  cond_br undef, left, right

left:
  %t = destructure_struct %v : $Box<T>
  destroy_value %t : $T
  br left2

left2:
  %agg2 = apply undef<Box<Box<T>>>() : $@convention(thin) <T> () -> (@out T)
  br merge(%agg2 : $Box<Box<T>>)

right:
  %agg = struct $Box<Box<T>>(%v : $Box<T>)
  br merge(%agg : $Box<Box<T>>)

other:
  destroy_value %v : $Box<T>
  br other2

other2:
  %agg3 = apply undef<Box<Box<T>>>() : $@convention(thin) <T> () -> (@out T)
  br merge(%agg3 : $Box<Box<T>>)

merge(%p : @owned $Box<Box<T>>):
  destroy_value %p : $Box<Box<T>>
  br exit

exit:
  %retval = tuple ()
  return %retval : $()
}
```

This is now `f190_vexing_incoming_def_projection` in `address_lowering_phi.sil`.